harness/determined/core/_metrics.py

```diff
@@ -0,0 +1,168 @@
+import logging
+import queue
+import threading
+from typing import Any, Dict, List, Optional
+
+from determined.common import api
+from determined.common.api import bindings
+
+logger = logging.getLogger("determined.core")
+
+
+class MetricsContext:
+    """Gives access to metrics reporting during trial tasks.
+
+    Metrics reported to ``MetricsContext`` are published to a queue, which is consumed by a
+    background thread that reports them to the master.
+    """
+
+    def __init__(
+        self,
+        session: api.Session,
+        trial_id: int,
+        run_id: int,
+    ) -> None:
+        self._session = session
+        self._trial_id = trial_id
+        self._run_id = run_id
+
+        self._error_queue = queue.Queue()
+        self._shipper = _Shipper(
+            session=self._session,
+            trial_id=self._trial_id,
+            run_id=self._run_id,
+            error_queue=self._error_queue,
+        )
+
+    def report(
+        self,
+        group: str,
+        steps_completed: int,
+        metrics: Dict[str, Any],
+        batch_metrics: Optional[List[Dict[str, Any]]] = None,
+    ) -> None:
+        # Check for thread exceptions here since we're not polling.
+        if not self._error_queue.empty():
+            err_msg = self._error_queue.get(block=False)
+            logger.error(f"Error reporting metrics: {err_msg}")
+            raise err_msg
+
+        self._shipper.publish_metrics(
+            group=group,
+            steps_completed=steps_completed,
+            metrics=metrics,
+            batch_metrics=batch_metrics,
+        )
+
+    def start(self) -> None:
+        self._shipper.start()
+
+    def close(self) -> None:
+        self._shipper.stop()
+        self._shipper.join()
+
+
+class _TrialMetrics:
+    def __init__(
+        self,
+        group: str,
+        steps_completed: int,
+        metrics: Dict[str, Any],
+        batch_metrics: Optional[List[Dict[str, Any]]] = None,
+    ):
+        self.group = group
+        self.steps_completed = steps_completed
+        self.metrics = metrics
+        self.batch_metrics = batch_metrics
```

Comment:
don't we have something like this class elsewhere?

Reply:
we have `determined.common.experimental.metrics.TrialMetrics` but it's under the experimental namespace. 
the profiler `determined.core._profiler` had a similar object for the same purpose, but i've refactored profiler to now use the MetricsContext